### PR TITLE
Support for SANSv2 autoprocess in DrILL

### PR DESCRIPTION
**Pull request by gui-co**
**Description of work.**

This PR add the support for the new ILL SANS reduction algorithm (see #32287).

This algorithm is now able to process kinetics data. In this mode, several rows (from the same group) are combined and submitted to a single call to the algorithm.

**To test:**

Use these data: [data.zip](https://github.com/mantidproject/mantid/files/8350876/data.zip)
and these masks: [masks.zip](https://github.com/mantidproject/mantid/files/8368888/masks.zip)

The python script has been provided to the users in order to test the new SANS algorithm. The following rundex corresponds to its "translation" for the DrILL interface.

Open the interface (Interface -> ILL -> DrILL) and load the rundex (Ctrl-O). Then process the table and compare the output with the one coming from the python script.

****python script****
```python
from mantid.simpleapi import SANSILLMultiProcess                                
                                                                                
cads        = ['17285', '17284', '17082']                                       
trbeam      = '17097'                                                           
emptybeams  = ['17112', '17097', '17098']                                       
cans        = ['17109', '17095', '17099']                                       
ctr         = '17096'                                                           
masks       = ['mask3m', 'mask_8m', 'mask_34m']                                 
flat_fields = ['flat_field_3m', 'flat_field_8m', 'flat_field_8m']               
edge        = 'edge_mask'                                                       
                                                                                
sampleTr  = ['17162', '17166', '17166', '17162']                                
                                                                                
samplesD1 = ['17251', '17177', '17257', '17202']                                
samplesD2 = ['17192', '17194', '17200', '17196+17198']                          
samplesD3 = ['17180+17184+17188', '17182+17186+17190', '17208+17212',           
             '17206+17218']                                                     
                                                                                
SANSILLMultiProcess(SampleRunsD1=','.join(samplesD1),                           
                    SampleRunsD2=','.join(samplesD2),                           
                    SampleRunsD3=','.join(samplesD3),                           
                    DarkCurrentRuns=','.join(cads),                             
                    EmptyBeamRuns=','.join(emptybeams),                         
                    EmptyContainerRuns=','.join(cans),                          
                    SampleTrRunsW1=','.join(sampleTr),                          
                    ContainerTrRuns=ctr,                                        
                    TrEmptyBeamRuns=trbeam,                                     
                    TrDarkCurrentRuns=cads[1],                                  
                    DefaultMask=edge,                                           
                    BeamStopMasks=','.join(masks),                              
                    FlatFields=','.join(flat_fields),                           
                    NormaliseBy='Time',                                         
                    CalculateResolution='DirectBeam',                           
                    BinningFactor=2,                                            
                    OutputWorkspace='script')
```
****rundex file****
```json
{
    "Instrument": "D11",
    "AcquisitionMode": "SANS v2",
    "VisualSettings": {
        "FoldedColumns": [
            "SampleRunsD4",
            "SampleRunsD5",
            "FluxRuns",
            "SampleTrRunsW2",
            "Solvents",
            "SampleThickness",
            "SampleNames"
        ],
        "HiddenColumns": [],
        "ColumnsOrder": [
            "SampleRunsD1",
            "SampleRunsD2",
            "SampleRunsD3",
            "SampleRunsD4",
            "SampleRunsD5",
            "DarkCurrentRuns",
            "EmptyBeamRuns",
            "FluxRuns",
            "EmptyContainerRuns",
            "SampleTrRunsW1",
            "SampleTrRunsW2",
            "TrDarkCurrentRuns",
            "ContainerTrRuns",
            "TrEmptyBeamRuns",
            "BeamStopMasks",
            "FlatFields",
            "Solvents",
            "SampleThickness",
            "SampleNames",
            "OutputWorkspace"
        ]
    },
    "GlobalSettings": {
        "SampleRunsD1": "",
        "SampleRunsD2": "",
        "SampleRunsD3": "",
        "SampleRunsD4": "",
        "SampleRunsD5": "",
        "DarkCurrentRuns": "",
        "EmptyBeamRuns": "",
        "FluxRuns": "",
        "EmptyContainerRuns": "",
        "SampleTrRunsW1": "",
        "SampleTrRunsW2": "",
        "TrDarkCurrentRuns": "",
        "ContainerTrRuns": "",
        "TrEmptyBeamRuns": "",
        "SensitivityMap": "",
        "DefaultMask": "edge_mask",
        "BeamStopMasks": "",
        "FlatFields": "",
        "Solvents": "",
        "TransmissionThetaDependent": true,
        "NormaliseBy": "Time",
        "TrBeamRadius": [
            0.1
        ],
        "BeamRadius": [
            0.25
        ],
        "SampleThickness": [
            0.1
        ],
        "SampleThicknessFrom": "User",
        "SampleNames": [],
        "SampleNamesFrom": "RunNumber",
        "WaterCrossSection": 1.0,
        "ProduceSensitivity": false,
        "SensitivityWithOffsets": false,
        "OutputType": "I(Q)",
        "DistancesAtWavelength2": [],
        "SubAlgorithmOffset": true,
        "OutputBinning": "",
        "CalculateResolution": "DirectBeam",
        "DefaultQBinning": "PixelSizeBased",
        "BinningFactor": "2",
        "NumberOfWedges": 0,
        "WedgeAngle": 30.0,
        "WedgeOffset": 0.0,
        "AsymmetricWedges": false,
        "WavelengthRange": [
            1.0,
            10.0
        ],
        "ShapeTable": "",
        "OutputPanels": false,
        "PerformStitching": true,
        "ManualScaleFactors": [],
        "TieScaleFactors": false,
        "ScaleFactorCalculation": "MedianOfRatios",
        "StitchReferenceIndex": 1,
        "OutputWorkspace": ""
    },
    "ExportAlgorithms": [],
    "Samples": [
        {
            "SampleRunsD1": "17251",
            "SampleRunsD2": "17192",
            "SampleRunsD3": "17180+17184+17188",
            "DarkCurrentRuns": "17285,17284,17082",
            "EmptyBeamRuns": "17112,17097,17098",
            "EmptyContainerRuns": "17109,17095,17099",
            "SampleTrRunsW1": "17162",
            "ContainerTrRuns": "17096",
            "TrEmptyBeamRuns": "17097",
            "BeamStopMasks": "mask3m,mask_8m,mask_34m",
            "FlatFields": "flat_field_3m,flat_field_8m,flat_field_8m",
            "OutputWorkspace": "drill",
            "TrDarkCurrentRuns": "17284"
        },
        {
            "SampleRunsD1": "17177",
            "SampleRunsD2": "17194",
            "SampleRunsD3": "17182+17186+17190",
            "SampleTrRunsW1": "17166"
        },
        {
            "SampleRunsD1": "17257",
            "SampleRunsD2": "17200",
            "SampleRunsD3": "17208+17212",
            "SampleTrRunsW1": "17166"
        },
        {
            "SampleRunsD1": "17202",
            "SampleRunsD2": "17196+17198",
            "SampleRunsD3": "17206+17218",
            "SampleTrRunsW1": "17162"
        }
    ],
    "SamplesGroups": {
        "A": [
            0,
            1,
            2,
            3
        ]
    },
    "MasterSamples": {
        "A": 0
    }
}
```

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by darominski — CHANGES_REQUESTED** [review]
This looks generally good. I have some comments and questions, and these are dispersed throughout the code. I am still to test the consistency due to issue with masks.

The lack of masks fortunately revealed an issue in the `validateInputs` of the `SANSILLMultiProcess`: in L392, the `RuntimeError` `re` should be assigned to `issues` dictionary as a string.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/drill/model/DrillExportModel.py [hunk 1]
```diff
@@ -158,17 +156,13 @@ def _validCriteria(self, ws, algo):
         except:
             return False
 
-    def _onTaskSuccess(self, name):
+    def _onTaskSuccess(self, wsName, filename):
         """
         Triggered when the export finished with success.
 
         Args:
```
[the remark this answers is not in the document]
Comment (gui-co) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/drill/model/DrillExportModel.py:163]: Done.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/drill/model/DrillExportModel.py [hunk 2]
```diff
@@ -179,18 +173,14 @@ def _onTaskSuccess(self, name):
                 del self._exports[wsName]
                 self._logSuccessExport(wsName)
 
-    def _onTaskError(self, name, msg):
+    def _onTaskError(self, wsName, filename, msg):
         """
         Triggered when the export failed.
 
         Args:
```
[the remark this answers is not in the document]
Comment (gui-co) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/drill/model/DrillExportModel.py:181]: Done.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/drill/model/DrillTask.py [hunk 3]
```diff
@@ -83,28 +118,43 @@ def run(self):
...
+        self.observer.signals.finished.connect(self._onFinished)
         self.observer.observeProgress(self.alg)
         self.observer.signals.progress.connect(
                 lambda p : self.signals.progress.emit(self, p)
                 )
         try:
             ret = self.alg.execute()
             if not ret:
-                self.signals.finished.emit(self, 1, "")
+                self._onFinished(1, "")
```
[the remark this answers is not in the document]
Comment (gui-co) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/drill/model/DrillTask.py:135]: When the return value of the algorithm is `false`, the observer already logged the error message.

**Review by darominski — APPROVED** [review]
Thank you for the changes, LGTM. The provided test works as advertised, and the output is exactly the same in both cases. However, there are some warnings with saving the output after DrILL processing, with `SaveCanSAS1D`, the units are not compatible with all outputs of the group to be saved, but that's perhaps beyond the scope of this PR.

**Comment by gui-co** [comment]
Thanks, I will create a separate PR concerning the error handling in `SANSILLMultiProcess`.

**Pull request merged**